internal/locate/region_cache.go

```diff
@@ -2207,15 +2212,77 @@ func (c *RegionCache) batchScanRegions(bo *retry.Backoffer, keyRanges []pd.KeyRa
 
 		metrics.RegionCacheCounterWithBatchScanRegionsOK.Inc()
 		if len(regionsInfo) == 0 {
-			return nil, errors.Errorf(
+			backoffErr = errors.Errorf(
 				"PD returned no region, range num: %d, limit: %d",
 				len(keyRanges), limit,
 			)
+			continue
+		}
+		if regionsHaveGap(keyRanges, regionsInfo, limit) {
+			backoffErr = errors.Errorf(
+				"PD returned regions have gaps, range num: %d, limit: %d",
+				len(keyRanges), limit,
+			)
+			continue
 		}
 		return c.handleRegionInfos(bo, regionsInfo, opt.needRegionHasLeaderPeer)
 	}
 }
 
+func regionsHaveGap(ranges []pd.KeyRange, regionsInfo []*pd.Region, limit int) bool {
```

Comment:
we have a similar function in lightning/BR use case

https://github.com/pingcap/tidb/blob/1480bb127a9363e31e8b2e9f3808289ea54d9d28/br/pkg/streamhelper/regioniter.go#L103

Do you think it's better to let PD client do the retry rather than different callers? cc @okjiang

Reply:
We are kind to accept your feature request. Please feel free to file it to PD leader.

Reply:
Ok, I'll fix it in PD client.

Reply:
After discussion, we decides leave it as a TODO in the future.